src/main/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepExecution.java

```diff
@@ -215,8 +215,25 @@ public void stop(@NonNull Throwable cause) throws Exception {
     @Override public void onResume() {
         try {
             if (state == null) {
-                Run<?, ?> run = getContext().get(Run.class);
-                LOGGER.fine(() -> "No ExecutorStepDynamicContext found for node block in " + run + "; perhaps loading from a historical build record, hoping for the best");
+                var flowNode = getContext().get(FlowNode.class);
+                LOGGER.fine(() -> "node block " + getContext() + " not yet scheduled, checking for an existing queue item");
+                if (flowNode == null) {
+                    LOGGER.fine(() -> "No FlowNode found for node block " + getContext() + "; can't recover" );
+                } else {
+                    var action = flowNode.getAction(QueueItemActionImpl.class);
+                    if (action == null) {
+                        LOGGER.fine(() -> "No QueueItemAction found for node block " + getContext() + "; can't recover");
+                    } else {
+                        LOGGER.fine(() -> "QueueItemAction with id=" + action.id + " found for node block " + getContext());
+                        var queueItem = action.itemInQueue();
+                        if (queueItem == null) {
+                            LOGGER.fine(() -> "Could not find queue item " + action.id + ", rescheduling it");
+                            start();
```

Comment:
I thought of a minor problem with this: https://github.com/jenkinsci/workflow-durable-task-step-plugin/blob/6ae8123a901592f1c4ce1920d4968826a2d8d665/src/main/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepExecution.java#L139 will add a _second_ `QueueItemActionImpl`, which means that if the item is still waiting in queue when the controller next terminates (even gracefully), the next call to `onResume` will find the first action in https://github.com/jenkinsci/workflow-durable-task-step-plugin/blob/711657d1bde3b1acca492f53367dbe5441ebb4e4/src/main/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepExecution.java#L223 and so it will schedule a duplicate queue item.